File: kits/python/conformance.py
"""The Python kit answering the conformance subject contract.

Kit-specific glue: seven verbs over JSON lines, a warden stood up from handed
keys, a door handed bytes, and the records read back as Article IX's ``cargo``.
It is written from ``papers/quo-conformance-contract.md`` and from this kit's own
public API, and it decides nothing: it stands a warden, hands it what it is
given, and reports what came back.

This file is an entry point, not part of the kit. It sits outside the package
so that importing ``quo`` still pulls in no host.

Three things this kit spells differently from the two the contract was written
beside, none of which is a decision this file is allowed to make:

- **It composes no cargo.** There is no ``pack``, so ``state`` reads the two
  records directly. The JS subject already does exactly this; only the Go one
  goes through a packer, so the contract has never required one.
- **Its judgment raises rather than returns.** ``Silence`` is this kit's way of
  saying "no answer", so it is caught here and reported as ``null``. **Nothing
  else is caught at the door**, because a subject that swallowed every
  exception would turn a crash into a silence, and telling those two apart is
  the whole of I-2 and of the runner's own liveness check.
- **It arms nothing before a receive.** Its randomness is a ``mint`` callable
  handed to the constructor, so ``expecting``'s keys are served by seeding that
  rather than by an arming call.
"""

from __future__ import annotations

import json
import sys
from typing import Any, Callable, Optional

sys.path.insert(0, "src")

from quo import arithmetic, notation, warden as kit  # noqa: E402


def un(text: str) -> bytes:
    """Hex in, bytes out."""
    return bytes.fromhex(text)


def uno(text: Optional[str]) -> Optional[bytes]:
    """The same, where absent stays absent."""
    return None if text is None else bytes.fromhex(text)


def hx(raw: bytes) -> str:
    return bytes(raw).hex()


def hxo(raw: Optional[bytes]) -> Optional[str]:
    return None if raw is None else bytes(raw).hex()


class Queue:
    """A finite list drawn in order.

    Drawing past the end is a fault the scenario must hear about rather than a
    silent refill, because a kit that drew more than it was given has told us
    something.
    """

    def __init__(self, name: str, values: list) -> None:
        self.name = name
        self.values = list(values)
        self.at = 0

    def draw(self):
        if self.at >= len(self.values):
            raise RuntimeError(
                f"the {self.name} queue ran out after {len(self.values)}"
            )
        one = self.values[self.at]
        self.at += 1
        return one


class House:
    """Everything one stood-up door holds, so a fresh `stand` starts clean."""

    def __init__(self) -> None:
        self.warden: Optional[kit.Warden] = None
        self.clock: Optional[Queue] = None
        self.random: Optional[Queue] = None
        # The keys this door will mint for a being it is about to take in.
        # This kit draws them from `mint` like any other randomness, so they are
        # handed out ahead of the queue rather than through an arming call.
        self.expecting: list = []
        # Every ask this warden composed while judging the message in hand.
        self.onward: list = []


house = House()


def door_of() -> kit.Warden:
    """The warden this run stood up.

    Every verb but `stand` is asked after one, so a missing warden is a fault
    in the run rather than a state to branch on.
    """
    if house.warden is None:
        raise RuntimeError("no warden has been stood up")
    return house.warden


def queue_of(which: str) -> Queue:
    got = house.clock if which == "clock" else house.random
    if got is None:
        raise RuntimeError(f"no {which} queue has been handed in")
    return got


def mint() -> bytes:
    """The one source this kit draws every key from.

    **This is the place the contract and this kit disagree about shape.** The
    contract hands `expecting`'s two keys in separately from the `random`
    queue, because the kits it was written beside take them by an arming call.
    Here both come off one callable, so the arriving being's keys are handed
    out first and the queue serves everything after.

    The consequence is a constraint rather than a decision: it holds only while
    nothing seals an answer before the cargo arrives, because an answer's
    ephemeral is drawn from the same callable. Where that does not hold, the
    scenario diverges and the divergence is the finding.
    """
    if house.expecting:
        return house.expecting.pop(0)
    return queue_of("random").draw()


def being_of(one: dict) -> Callable[[str, bytes, kit.Leash], Optional[bytes]]:
    """The one thing a being in this contract ever does.

    A warden never makes an onward ask of its own: it hands the leash to the
    being it routed to and the being decides. So a being that calls out is the
    only way Article VIII's onward rules can be reached at all.

    **It decides nothing.** The scenario named the far warden, the being, the
    method and the ephemeral key, and what this returns is never asserted.

    **The leash spent is the one this kit handed the being**, passed straight
    back to `ask`, which reads the allowance off it at the moment of sealing.
    Recomputing it here would be the subject doing the arithmetic the case is
    about, and the case would then measure this file rather than the warden.
    """
    spec = one.get("onward")

    def invoke(name: str, args: bytes, leash: kit.Leash) -> Optional[bytes]:
        if not spec or spec["when"] != name:
            return b""
        row = door_of().relation(un(spec["at"]))
        if row is None:
            raise RuntimeError(f"no relation at {spec['at']}")
        method = spec.get("method")
        try:
            composed, _ = door_of().ask(
                row,
                un(spec["ephemeral"]),
                being=uno(spec.get("being")),
                method=(
                    {"name": method["name"], "args": un(method.get("args") or "")}
                    if method
                    else None
                ),
                seq=int(spec["seq"]),
                leash=leash,
            )
        except kit.Silence:
            # A leash with nothing left to spend composes nothing, and the being
            # answers anyway: Article VIII withholds the onward ask while "the
            # work already routed stands".
            return b""
        house.onward.append(composed)
        return b""

    return invoke


def cargo_of(being_pk: bytes) -> Optional[dict]:
    """Article IX's own shape, read off the records.

    Not a format this harness invented: `cargo` is what a warden's state looks
    like when it crosses, which is the level a check has to work at.
    """
    w = door_of()
    being = w.beings.get(being_pk)
    if being is None:
        return None
    standings = []
    for row in w.inbound:
        if being_pk not in [bytes(one) for one in row.beings]:
            continue
        standings.append(
            {
                "voice": hx(row.voice),
                "commitment": hx(row.commitment),
                # The name the commitment was minted under (Article XIV). This
                # kit calls it `minted_at`.
                "name": hx(row.minted_at),
                "beings": sorted(hx(one) for one in row.beings),
                "mark": str(row.mark or 0),
                "spent": [str(one) for one in sorted(row.spent)],
                "padlock": hxo(row.padlock),
                "hints": list(row.hints or ()),
            }
        )
    standings.sort(key=lambda one: str(one["voice"]))
    relations = []
    for row in w.outbound:
        relations.append(
            {
                "warden": hx(row.warden),
                "commitment": hx(row.commitment),
                "padlock": hx(row.padlock),
                "voice": hx(row.voice),
                "heir": hx(row.heir),
                "seq": str(row.seq or 0),
                # Two counters, never one field doing both (Article IX).
                "news": str(row.news or 0),
                "hints": list(row.hints or ()),
            }
        )
    relations.sort(key=lambda one: str(one["warden"]))
    return {
        "being": hx(being.pk),
        "digest": hx(being.digest),
        "cells": hx(being.cells or b""),
        "standings": standings,
        "relations": relations,
    }


def stand(order: dict) -> dict:
    house.__init__()
    spec = order["warden"]
    heir_secret = uno(spec.get("heirSeed"))
    house.warden = kit.Warden(
        name_secret=un(spec["nameSeed"]),
        padlock_secret=un(spec["padlockSeed"]),
        limit=int(spec.get("limit") or 0) or 1 << 62,
        mint=mint,
        heir=arithmetic.signing_public(heir_secret) if heir_secret else None,
    )
    w = door_of()
    w.hints = tuple(spec.get("hints") or ())
    house.clock = Queue("clock", [int(one) for one in order.get("clock") or []])
    house.random = Queue("random", [un(one) for one in order.get("random") or []])

    beings = []
    for one in order.get("beings") or []:
        pk = w.hold(
            one["blueprint"], being_of(one), un(one["seed"]), un(one["heirSeed"])
        )
        w.beings[pk].cells = un(one.get("cells") or "")
        beings.append(hx(pk))

    grants = []
    for one in order.get("grants") or []:
        heir = arithmetic.signing_public(un(one["heirSeed"]))
        # The kit's own arithmetic, called rather than reproduced. `invite`
        # would mint this too, but it also stamps the door's own padlock on the
        # row — and Article VII has the row keep "the padlock it named", the
        # voice's, which at grant time is nothing.
        commitment = arithmetic.commitment(w.name, heir)
        w.grant(
            arithmetic.signing_public(un(one["voiceSeed"])),
            commitment,
            [un(one["being"])],
            padlock=uno(one.get("padlock")),
            hints=one.get("hints") or (),
        )
        grants.append(
            {
                "warden": hx(w.name),
                "commitment": hx(commitment),
                "padlock": hx(w.padlock),
                "heir": hx(heir),
            }
        )

    for one in order.get("relations") or []:
        w.outbound.append(
            kit.Relation(
                warden=un(one["warden"]),
                commitment=un(one["commitment"]),
                padlock=un(one["padlock"]),
                voice=arithmetic.signing_public(un(one["voiceSeed"])),
                secret=un(one["voiceSeed"]),
                heir=arithmetic.signing_public(un(one["heirSeed"])),
                heir_secret=un(one["heirSeed"]),
                hints=tuple(one.get("hints") or ()),
            )
        )

    arriving = order.get("expecting")
    if arriving:
        house.expecting = [un(arriving["seed"]), un(arriving["heirSeed"])]
        text = notation.render(notation.parse(arriving["blueprint"]))
        w.blueprints[notation.digest(text)] = text

    for one in order.get("moved") or []:
        word = one.get("word") or {}
        w.pointers[un(one["being"])] = {
            "being": uno(word.get("being")),
            "successor": uno(word.get("successor")),
            "commitment": uno(word.get("commitment")),
            "name": uno(word.get("name")),
            "padlock": uno(word.get("padlock")),
            "hints": list(word.get("hints") or ()),
        }

    return {
        "warden": {"name": hx(w.name), "padlock": hx(w.padlock)},
        "beings": beings,
        "grants": grants,
    }


def door(order: dict) -> dict:
    """Bytes in, bytes out, or nothing — and nothing is silence.

    `onward` is every ask this warden composed while judging these bytes, in
    the order it composed them, and it is empty for every call that reached no
    being with an `onward` of its own.
    """
    house.onward = []
    try:
        judged = door_of().judge(un(order["bytes"]), clock=queue_of("clock").draw)
    except kit.Silence:
        return {"answer": None, "onward": [hx(one) for one in house.onward]}
    return {"answer": hx(judged.answer), "onward": [hx(one) for one in house.onward]}


def amend(order: dict) -> dict:
    """The house changing its own mind about a standing.

    This kit spells it as one call taking the row's whole membership, where the
    JS one takes an add and a remove list — so the two are computed here and
    the effect is the same. Add first, then remove, and taking the last being
    away is release, which this kit does for itself.
    """
    w = door_of()
    voice = un(order["voice"])
    row = next((one for one in w.inbound if bytes(one.voice) == voice), None)
    if row is None:
        raise kit.Silence("no such standing")
    beings = [bytes(one) for one in row.beings]
    for one in order.get("add") or []:
        if un(one) not in beings:
            beings.append(un(one))
    for one in order.get("remove") or []:
        if un(one) in beings:
            beings.remove(un(one))
    w.amend(voice, beings)
    return {}


def succeed(order: dict) -> dict:
    """The house moving its own name.

    This kit is one of the two that never sees the next heir's key, so it takes
    the commitment and leaves the seed beside it alone.
    """
    door_of().succeed(un(order["nameSeed"]), un(order["heirCommitment"]))
    return {}


def told(word, secret: bytes, peers: list, spec: list) -> list:
    """One piece of news per peer, each sealed with the key it was handed.

    A peer that left no way back composes nothing, which the kit decides and
    this only passes on.

    **The word is the kit's.** `depart` and `landed` compose it; this file says
    only which being left and where it went. Every field of it is reachable
    from here, so a subject that built it and asked the kit only to seal would
    go green while asserting nothing about the warden.
    """
    out = []
    for at, peer in enumerate(peers):
        if at >= len(spec):
            break
        one = spec[at]
        try:
            sealed = door_of().news(
                peer=peer,
                voice_secret=secret,
                word=word,
                seq=int(one["seq"]),
                ephemeral_secret=un(one["ephemeral"]),
                allowance={
                    "time": int(one["allowance"]["time"]),
                    "hops": int(one["allowance"]["hops"]),
                },
            )
        except kit.Silence:
            continue
        out.append(hx(sealed))
    return out


def depart(order: dict) -> dict:
    gone = order["gone"]
    word, secret, peers = door_of().depart(
        un(order["being"]),
        commitment=un(order["commitment"]),
        name=un(gone["name"]),
        padlock=un(gone["padlock"]),
        hints=gone.get("hints") or [],
    )
    return {"news": told(word, secret, peers, order.get("news") or [])}


def landed(order: dict) -> dict:
    word, secret, peers = door_of().landed(order.get("hints") or [])
    return {"news": told(word, secret, peers, order.get("news") or [])}


def state(order: dict) -> dict:
    """The records, as cargo, and the facts this kit cannot report at all.

    `relations` is every outbound row this door holds rather than the ones that
    would travel with this being: **this kit's `Relation` carries no holder**,
    so which being may spend which relation is a fact it has no way to give.
    Every scenario driven so far holds at most one outbound row, where the two
    lists are the same list; a door holding more would have to declare it in
    `cannot` rather than report a wrong one.
    """
    return {"cargo": cargo_of(un(order["being"])), "cannot": []}


def send(order: dict) -> dict:
    w = door_of()
    ask = order["ask"]
    row = w.relation(un(ask["at"]))
    if row is None:
        return {"error": f"no relation at {ask['at']}"}
    method = ask.get("method")
    try:
        composed, _ = w.ask(
            row,
            queue_of("random").draw(),
            being=uno(ask.get("being")),
            method=(
                {"name": method["name"], "args": un(method.get("args") or "")}
                if method
                else None
            ),
            allowance={
                "time": int(ask["allowance"]["time"]),
                "hops": int(ask["allowance"]["hops"]),
            },
            seq=int(ask["seq"]),
        )
    except kit.Silence:
        # A refusal to send is an ordinary expected outcome, not an error.
        return {"bytes": None}
    return {"bytes": hx(composed)}


def read(order: dict) -> dict:
    try:
        answer = door_of().hear(un(order["answer"]))
    except kit.Silence:
        return {"answer": None}
    data = answer.get("data")
    return {
        "answer": {
            "warden": hx(answer["warden"]),
            "seq": str(answer["seq"]),
            "data": hxo(data),
        }
    }


VERBS: dict[str, Callable[[dict], dict]] = {
    "stand": stand,
    "door": door,
    "amend": amend,
    "succeed": succeed,
    "state": state,
    "send": send,
    "read": read,
    "depart": depart,
    "landed": landed,
}


def obey(order: dict) -> Any:
    verb = VERBS.get(order.get("do", ""))
    if verb is None:
        return {"error": f"no such verb: {order.get('do')}"}
    try:
        return verb(order)
    except Exception as thrown:  # noqa: BLE001 - reported, never swallowed
        return {"error": str(thrown) or thrown.__class__.__name__}


def main() -> None:
    for line in sys.stdin:
        if not line.strip():
            continue
        sys.stdout.write(json.dumps(obey(json.loads(line))) + "\n")
        sys.stdout.flush()


if __name__ == "__main__":
    main()

File: kits/js/cmd/conformance.js
// The JS kit answering the subject contract, and nothing more. A subject is
// kit-specific glue: nine verbs over JSON lines, a warden stood up from handed
// keys, a door handed bytes, and the records read back as Article IX's `cargo`.
// Every other kit writes its own, and a stranger's is the only thing they must
// implement to be driven.
//
// What this file must never do is decide anything. It stands a warden, hands
// it what it is given, and reports what came back. The judgment is the kit's;
// the expectations are the scenario's; this is the wire between them.
//
// A subject is not a host, and it stands below that seam on purpose. Proving
// the kit from outside means composing what no application may: an ask naming
// neither being nor method, argument bytes deliberately malformed, an ask at a
// being whose blueprint it does not hold, the seq it spent read back directly.
// The host's surface refuses every one of those by design — a handle encodes
// through the blueprint, so it can never produce the malformed input a refusal
// is asserted with. So this file drives the warden's own entry point rather
// than the kit's host module, and the seam never grows a raw-ask surface to
// accommodate it: that would ship every application a public way around the
// blueprint, permanently, for the benefit of the harness.
import { Warden, current, depart, landed, news, signingPair } from '../src/index.js';

const hex = (bytes) => Buffer.from(bytes).toString('hex');
const un = (text) => Uint8Array.from(Buffer.from(text, 'hex'));

// The two Article II freedoms, handed in as queues. A finite list, drawn in
// order: the whole reason a scenario's bytes can be pinned at all. Drawing past
// the end is a fault the scenario must hear about rather than a silent refill,
// because a kit that drew more than it was given has told us something.
function queue(name, values) {
  let at = 0;
  return () => {
    if (at >= values.length) throw new Error(`the ${name} queue ran out after ${values.length}`);
    const one = values[at];
    at += 1;
    return one;
  };
}

let warden = null;
let clock = null;
let random = null;
// Every ask this warden composed while judging the message in hand. A door
// with no being that calls out leaves it empty, which is the same thing as
// nothing having been handed onward.
let onward = [];

// The one thing a being in this contract does. A warden never makes an onward
// ask of its own — it hands the leash to the being it routed to — so a being
// that calls out is the only way Article VIII's onward rules can be reached at
// all. It decides nothing: the scenario named the far warden, the being, the
// method and the ephemeral key, and what this returns is never asserted.
function being(one) {
  if (!one.onward) return {};
  const spec = one.onward;
  return {
    [spec.when]: async () => {
      const row = warden.outbound.find((at) => hex(at.warden) === spec.at);
      if (!row) throw new Error(`no relation at ${spec.at}`);
      // The leash the kit handed this call, spent as it was handed. Recomputing
      // it here would be the subject doing the arithmetic the case is about.
      const leash = current()?.leash ?? null;
      const composed = await warden.ask(row, {
        seq: BigInt(spec.seq),
        leash,
        being: spec.being ? un(spec.being) : null,
        method: spec.method ? { name: spec.method.name, args: un(spec.method.args ?? '') } : null,
        random: un(spec.ephemeral),
      });
      // A leash with nothing left to spend composes nothing, and the being
      // answers anyway: Article VIII has the onward ask withheld while "the
      // work already routed stands". A being that failed here would be turning
      // its own kit's refusal into the door's silence.
      if (composed) onward.push(composed);
      // What this being answers belongs to its blueprint and is asserted
      // nowhere: the scenario is about what went onward.
      return 0n;
    },
  };
}

// Article IX's own shape, read off the records. Not a format this harness
// invented: `cargo` is what a warden's state looks like when it crosses, which
// is the level a check has to work at, because five kits store differently and
// all five must be readable the same way.
function cargoOf(beingPk) {
  const being = warden.beings.get(hex(beingPk));
  if (!being) return null;
  const at = hex(beingPk);
  const standings = [];
  for (const row of warden.inbound.values()) {
    if (!row.beings.has(at)) continue;
    standings.push({
      voice: hex(row.voice),
      commitment: hex(row.commitment),
      // The name the commitment was minted under (Article XIV).
      name: hex(row.name),
      beings: [...row.beings].sort(),
      mark: (row.mark ?? 0n).toString(),
      spent: row.window().map((one) => one.toString()),
      padlock: row.padlock ? hex(row.padlock) : null,
      hints: row.hints ?? [],
    });
  }
  // Ordered so two readings of one state are one text. The law derives an
  // estate's order (Article X) and says nothing about a cargo's; this order is
  // the harness's, and it decides nothing about the protocol.
  standings.sort((a, b) => (a.voice < b.voice ? -1 : a.voice > b.voice ? 1 : 0));
  const relations = (warden.relationsOf ? warden.relationsOf(beingPk) : []).map((row) => ({
    warden: hex(row.warden),
    commitment: hex(row.commitment),
    padlock: hex(row.padlock),
    voice: hex(row.voice.pk),
    heir: hex(row.heir.pk),
    seq: (row.seq ?? 0n).toString(),
    // Two counters, never one field doing both (Article IX). The mark kept for
    // that far warden's news lives on the row's own marks, which is where the
    // kit's `pack` reads it from.
    news: (row.marks?.mark ?? 0n).toString(),
    hints: row.hints ?? [],
  }));
  return {
    being: hex(being.pk),
    digest: hex(being.digest),
    cells: hex(being.cells()),
    standings,
    relations,
  };
}

// One piece of news per peer, each with the key it seals with and the number it
// spends against that peer's own mark — both handed in, like every other draw.
// A peer that left no way back composes nothing, which the kit decides and this
// file only passes on.
async function told(word, voice, rows, spec) {
  const out = [];
  for (let at = 0; at < rows.length; at += 1) {
    const one = spec[at];
    if (!one) break;
    const sealed = news(warden, {
      peer: rows[at],
      voice,
      word,
      seq: BigInt(one.seq),
      allowance: { time: BigInt(one.allowance.time), hops: BigInt(one.allowance.hops) },
      random: un(one.ephemeral),
    });
    if (sealed) out.push(hex(await sealed));
  }
  return out;
}

const verbs = {
  // Stand a warden up from handed keys, hold the beings, write the standings,
  // and take the two queues. Everything after this is deterministic.
  async stand(order) {
    warden = await Warden.open({
      nameSeed: un(order.warden.nameSeed),
      padlockSeed: un(order.warden.padlockSeed),
      heirSeed: un(order.warden.heirSeed),
      limit: BigInt(order.warden.limit ?? 0),
      hints: order.warden.hints ?? [],
      // The two Article II freedoms, handed to the warden at the moment it
      // opens: it draws from the scenario's queues and never reaches for
      // either.
      clock: () => clock(),
      random: () => random(),
    });
    const beings = [];
    for (const one of order.beings ?? []) {
      const { being: pk } = await warden.hold(being(one), {
        seed: un(one.seed),
        heirSeed: un(one.heirSeed),
        blueprint: one.blueprint,
        cells: () => un(one.cells ?? ''),
      });
      beings.push(hex(pk));
    }
    const grants = [];
    for (const one of order.grants ?? []) {
      const invitation = await warden.grant(un(one.being), {
        voiceSeed: un(one.voiceSeed),
        heirSeed: un(one.heirSeed),
        padlock: one.padlock ? un(one.padlock) : null,
        hints: one.hints ?? [],
      });
      grants.push({
        warden: hex(invitation.warden),
        commitment: hex(invitation.commitment),
        padlock: hex(invitation.padlock),
        heir: hex(invitation.heirPublic),
      });
    }
    // The outbound rows: invitations this ground holds at other houses, each
    // held by the being that may spend it.
    for (const one of order.relations ?? []) {
      const voice = await signingPair(un(one.voiceSeed));
      const heir = await signingPair(un(one.heirSeed));
      warden.remember(
        {
          warden: un(one.warden),
          commitment: un(one.commitment),
          padlock: un(one.padlock),
          heirPublic: heir.pk,
          heirSecret: heir.secret,
          hints: one.hints ?? [],
        },
        {
          voicePk: voice.pk,
          voiceSecret: voice.secret,
          being: one.being ? un(one.being) : null,
        },
      );
    }
    // The being this door is about to take in. Article IX's `receive` arrives
    // as an ordinary ask, so nothing else is needed to drive it — but a door
    // that had not been told is a door any holder could push a being into.
    if (order.expecting) {
      await warden.expect({
        seed: un(order.expecting.seed),
        heirSeed: un(order.expecting.heirSeed),
        object: {},
        blueprint: order.expecting.blueprint,
        cells: () => un(order.expecting.cells ?? ''),
      });
    }
    // The beings that have gone, and the succession this door published for
    // each. An absent field of the `word` stays absent.
    for (const one of order.moved ?? []) {
      const word = one.word ?? {};
      const key = (name) => (word[name] ? un(word[name]) : null);
      warden.point(un(one.being), {
        being: key('being'),
        successor: key('successor'),
        commitment: key('commitment'),
        name: key('name'),
        padlock: key('padlock'),
        hints: word.hints ?? [],
      });
    }
    clock = queue(
      'clock',
      (order.clock ?? []).map((one) => BigInt(one)),
    );
    random = queue(
      'random',
      (order.random ?? []).map((one) => un(one)),
    );
    return {
      warden: { name: hex(warden.name.pk), padlock: hex(warden.padlock.pk) },
      beings,
      grants,
    };
  },

  // The door. Bytes in, bytes out, or nothing — and nothing is silence.
  async door(order) {
    // One draw per message, taken before the door is handed anything and spent
    // only if the judgment reaches step 8. This kit takes the draw as a value
    // and the clock as a function; another takes both at once. Both are
    // conforming, and it is exactly why no scenario asserts how far a queue was
    // drawn down.
    onward = [];
    const answer = await warden.judge(un(order.bytes), { clock, random: random() });
    return {
      answer: answer ? hex(answer) : null,
      onward: onward.map(hex),
    };
  },

  // The caller's half: one ask composed down an outbound row. `null` bytes are
  // a refusal to send, which Article III makes an ordinary outcome — the ask
  // never reaches the road — and never an error.
  async send(order) {
    const ask = order.ask;
    const row = warden.outbound.find((one) => hex(one.warden) === ask.at);
    if (!row) return { error: `no relation at ${ask.at}` };
    const bytes = await warden.ask(row, {
      seq: BigInt(ask.seq),
      commitment: ask.commitment ? un(ask.commitment) : null,
      allowance: { time: BigInt(ask.allowance.time), hops: BigInt(ask.allowance.hops) },
      being: ask.being ? un(ask.being) : null,
      method: ask.method ? { name: ask.method.name, args: un(ask.method.args ?? '') } : null,
      random: random(),
    });
    return { bytes: bytes ? hex(bytes) : null };
  },

  // And the caller's other half: one answer judged at this end. `null` is the
  // whole of what any failed check looks like, and the kit is asked which
  // warden the ask went to rather than reading it off the record it is judging.
  async read(order) {
    // The kit's own caller surface, which makes all four of Article XII's
    // checks: the two the envelope carries, that the warden is one this ground
    // holds a relation with, and that an ask is awaiting under that padlock,
    // that warden and that seq. `at` is not handed to it — this kit satisfies
    // the warden check from its own record, which is the stronger form.
    const answer = await warden.hear(un(order.answer));
    if (!answer) return { answer: null };
    return {
      answer: {
        warden: hex(answer.warden),
        seq: (answer.seq ?? 0n).toString(),
        data: answer.data === null || answer.data === undefined ? null : hex(answer.data),
      },
    };
  },

  // The records, as `cargo`, and the facts this kit cannot report at all.
  // The house changing its own mind about a standing. Nothing crosses, nothing
  // is spent, and this kit spells it as one call where another may spell it as
  // two — the contract fixes the effect and not the spelling.
  amend(order) {
    warden.amend(un(order.voice), {
      add: (order.add ?? []).map(un),
      remove: (order.remove ?? []).map(un),
    });
    return {};
  },

  // The house moving its own name. Like an amend it crosses no wire, and like
  // an amend the whole of what it claims is in the record afterwards and in
  // what the door answers to next. This kit takes the next heir as a seed; a
  // kit that never sees that key takes the commitment instead.
  //
  // A key this door never committed to comes back null here, and that is an
  // error rather than nothing: a scenario that believed it had succeeded a
  // door and had not would assert every exchange after it against a door that
  // never moved.
  async succeed(order) {
    const moved = await warden.succeed({
      nameSeed: un(order.nameSeed),
      heirSeed: un(order.heirSeed),
    });
    if (!moved) throw new Error('that key is not the heir this door committed to');
    return {};
  },

  // The origin's half of a migration's news. This kit holds the being's heir
  // itself, so the handed-in seed is ignored; a kit that does not hold it takes
  // that seed instead, and both name one key.
  //
  // The word is the kit's — `depart` composes it — and this file only says
  // which being left and where it went. A subject that built the word and asked
  // the kit to seal it would assert nothing about the warden.
  async depart(order) {
    const gone = depart(warden, un(order.being), {
      commitment: un(order.commitment),
      name: un(order.gone.name),
      padlock: un(order.gone.padlock),
      hints: order.gone.hints ?? [],
    });
    if (!gone) throw new Error('no being of that name');
    return { news: await told(gone.word, gone.voice, gone.peers, order.news ?? []) };
  },

  // The destination's half, after a cargo has arrived. It needs nothing but the
  // roads this door answers on — and this kit fixed those at `stand`, so even
  // those it takes from its own record.
  async landed(order) {
    const here = landed(warden);
    if (!here) throw new Error('nothing has arrived at this door');
    return { news: await told(here.word, here.voice, here.peers, order.news ?? []) };
  },

  // Declared rather than left to be inferred from a null, because a null value
  // and an unanswerable question are different things and only the subject
  // knows which is which. This kit now reports every Article IX field, so the
  // list is empty.
  state(order) {
    return { cargo: cargoOf(un(order.being)), cannot: [] };
  },
};

export async function obey(order) {
  const verb = verbs[order.do];
  if (!verb) return { error: `no such verb: ${order.do}` };
  try {
    return await verb(order);
  } catch (thrown) {
    return { error: String(thrown && thrown.message ? thrown.message : thrown) };
  }
}

// Spawned as a process, the contract is JSON lines on stdin and stdout. Driven
// in-process, `obey` is the same contract without the pipe.
if (process.argv[1] && process.argv[1].endsWith('conformance.js')) {
  let buffer = '';
  process.stdin.setEncoding('utf8');
  process.stdin.on('data', async (chunk) => {
    buffer += chunk;
    const lines = buffer.split('\n');
    buffer = lines.pop() ?? '';
    for (const line of lines) {
      if (!line.trim()) continue;
      const out = await obey(JSON.parse(line));
      process.stdout.write(`${JSON.stringify(out)}\n`);
    }
  });
}
